mitmproxy/proxy/layers/http/__init__.py

```diff
@@ -308,6 +309,25 @@ def send_response(self, already_streamed: bool = False):
         """We have either consumed the entire response from the server or the response was set by an addon."""
         assert self.flow.response
         self.flow.response.timestamp_end = time.time()
+
+        is_websocket = (
+            self.flow.response.status_code == 101
+            and
+            self.flow.response.headers.get("upgrade", "").lower() == "websocket"
+            and
+            self.flow.request.headers.get("Sec-WebSocket-Version", "") == "13"
+            and
+            self.context.options.websocket
+        )
+        if is_websocket:
+            # We need to set this before calling the response hook
+            # so that addons can determine if a WebSocket connection is following up.
+            self.flow.websocket = WebSocketData()
+            if self.flow.request.scheme == "http":
+                self.flow.request.scheme = "ws"
+            elif self.flow.request.scheme == "https":
+                self.flow.request.scheme = "wss"
```

Comment:
I have hard feelings about this. `flow` is an HTTPFlow that happens to have a WebSocket using the same connection. If this is needed for the UI then the UI layer needs to handle this logic. But the scheme should never change.

```py
def requestheaders(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('requestheaders')
      print(flow.request.scheme)

def request(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('request')
      print(flow.request.scheme)

def responseheaders(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('responseheaders')
      print(flow.request.scheme)

def response(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('response')
      print(flow.request.scheme)
```

```
requestheaders
https

request
https

responseheaders
https

response
wss
```

![giphy](https://user-images.githubusercontent.com/679144/110762949-a5c44f80-8251-11eb-894e-4cfba6b962b3.gif)

Reply:
Maybe a `flow.websocket.scheme` property?

Reply:
I don't think it's that bad, but I don't feel as strongly. I've changed it to the UI layer. 👍 
(`flow.websocket.scheme` is not really necessary)